proposals/reward-proposal-v2-reward-calculation.md

```diff
@@ -0,0 +1,109 @@
+reward payment without the ability for double-entry booking.
+
+Given these two drawbacks above, we are proposing an improved way to handle
+epoch reward - `V2`. The `V2` approach is based on `V1` but with the
+improvements to address these two drawbacks.
+
+To make the discussion clearer, the `V2` approach is split into two proposals:
+`rewards calculation` and `rewards distribution`. In this proposal, we will
+focus on "reward calculation". A follow-up proposal will talk about "reward
+distribution".
+
+## Proposed Solutions
+
+Unlike the current epoch rewards computation approach, in which computing the
+rewards happens at the first slot of epoch boundary in the bank runtime
+directly, the new reward computation is moved to a background service in the
+proposed approach.
+
+This is very similar to the `V1` proposal, a separate service,
+"EpochRewardCalculationService" is created at the start time of the validator.
+This service is listening on a channel for incoming rewards calculation
+requests, and handle the requests to compute the rewards.
+
+There are two main differences between the `V2` and `V1` approaches.
+
+The first difference is that instead of sharing the `StakeCache` with the
+runtime during reward computation, the `StakeCache` is cloned at slot-height 0
+in the `V2` epoch. In this way, we can avoid locking stake accounts and vote
+accounts to prevent updates during the reward calculation.
+
+And the reward computation can last for the next `N` block-height. At the end of
+`Nth` block-height, the reward computation result should be available. In case
+that the reward result is not available yet at the `Nth` block-height, the
+runtime will wait until the computation is finished before processing any blocks
+with height greater than `N`.
+
+The second difference is to introduce the the `reward proof`, which will enable
+double-entry booking capability for reward distributions. To make sure that we
+have a consistent proof of the epoch rewards across different validator nodes, a
+merkle tree is computed from the epoch rewards of all stake and vote accounts.
+The root of the merkle tree and total rewards will be recorded on the network in
+a new system account `EpochRewardProof` This `EpochRewardProof` account
+maintains a history of epoch rewards. The internal format of `EpochRewardProof`
+could be a vector of entries of `(epoch, (reward_amount, root_hash))`, where
+each entry corresponds to the reward for one particular epoch.
+
+`EpochRewardProof` will be stored at the end of the reward computation, i.e.
+`Nth` block-height of the epoch. `EpochRewardProof` will also be used later for
+epoch reward distribution. To facilitate the reward distribution, a family of
+`EpochRewardReserve` accounts, which holds the partial balance of the total
+rewards, will be replenished with the rewards to be distributed. See `Epoch
+Reward Distribution` proposal for more details.
+
+### Challenges and Difficulties
```

Comment:
I assume these are essentially outstanding TODO items that would be resolved before this proposal is merged?

Reply:
Yes, they are. 
And those items are the topics that we are not sure how to do it, need to brainstorm, discuss. In the end, we have to come up with mutually agreed solutions before merging.